design.md

```diff
@@ -0,0 +1,226 @@
+# Design proposal
+
+## Objective
+
+Implement a simple remote process supervisor that lets its users start, stop and monitor arbitrary commands, as well as retrieving their outputs as streams. A CLI tool will be provided to interact with the server via its gRPC-powered API.
+
+## Assumptions, decisions and tradeoffs
+- Given many different implementation options, the most straighforward one will be chosen unless further requirements are provided
+- The jobs provided by users are well-intentioned and not malicious, the resource control mechanisms described below act as a safeguard against user/software errors, not targeted attacks
+- There will not be any attempts to persist the jobs or recover them on failure
+- The output from the jobs will be held in memory until consumed or discarded
+- All the jobs get the same set of resource limits
+- Everything contained in this document is a proposal and subject to approval and improvements, the final code may not exactly match this document
+
+## Library
+
+The library will expose functions to start a given job, stop it, query its status and get a stream of its standard output and standard error. This is the proposed structure:
+
+```go
+func NewSupervisor() *Supervisor
+	Creates a new supervisor object that will handle the job operations.
+
+func (s *Supervisor) StartJob(cmd string, args ...string) (string, error)
+	Starts a new job with the provided command and arguments, using the os/exec standard package.
+
+func (s *Supervisor) StopJob(id string) (int, error)
+	Must be called to free the resources allocated by the supervisor for this process. If the process has not finished running, it will get killed with os.Process.Kill().
+
+func (s *Supervisor) JobStatus(id string) string
+	Returns a string with the status of the given job: "Started" or "Done".
+
+func (s *Supervisor) JobStdOut(id string) (io.ReadCloser, error)
+	Returns a stream that corresponds to the standard output of the process.
+
+func (s *Supervisor) JobStdErr(id string) (io.ReadCloser, error)
+	Returns a stream that corresponds to the standard error of the process.
+```
+
+## API
+
+The API will use the gRPC framework for communication and Protocol Buffers as the serialization format.
+
+### Protocol definition
+
+```protobuf
+syntax = "proto3";
+
+package overseer;
+
+message Job {
+    string command = 1;
+    repeated string arguments = 2;
+}
+
+message JobID {
+    string id = 1;
+}
+
+message StopResponse {
+    int64 exitCode = 1;
+}
+
+enum Status {
+    STARTED = 0;
+    DONE = 1;
+}
+
+message StatusResponse {
+    Status status = 1;
+}
+
+message OutputChunk {
+    bytes output = 1;
+}
+
+service JobworkerService {
+    rpc Start(Job) returns (JobID) {}
+    rpc Stop(JobID) returns (StopResponse) {}
+    rpc Status(JobID) returns (StatusResponse) {}
+    rpc StdOut(JobID) returns (stream OutputChunk) {}
+    rpc StdErr(JobID) returns (stream OutputChunk) {}
```

Comment:
This works, but feels like you can collapse them into one RPC, any reason why you have two?

Reply:
I mostly did it like that so the cli tool could have access to those streams separately and without mixing the standard error of my tool together with the standard error of remote job. To be clear, the cli tool will always write the remote job's output to the standard output (be it from stdout or stderr) and any errors (like authentication, network, etc) to the standard error. I think this way would be more script-friendly.